spiped/main.c

```diff
@@ -285,20 +292,28 @@ main(int argc, char * argv[])
 
 	/* Start accepting connections. */
 	if ((dispatch_cookie = dispatch_accept(s, opt_t, opt_R ? 0.0 : opt_r,
-	    sas_t, opt_d, opt_f, opt_g, opt_j, K, opt_n, opt_o)) == NULL) {
+	    sas_t, opt_d, opt_f, opt_g, opt_j, K, opt_n, opt_o,
+	    opt_1 ? &conndone : NULL)) == NULL) {
 		warnp("Failed to initialize connection acceptor");
 		goto err5;
 	}
 
 	/* Infinite loop handling events. */
-	do {
-		if (events_run()) {
-			warnp("Error running event loop");
-			goto err6;
-		}
-	} while (1);
+	if (events_spin(&conndone)) {
+		warnp("Error running event loop");
+		goto err6;
+	}
+
+	/* Clean up. */
+	dispatch_shutdown(dispatch_cookie);
```

Comment:
In non-error-handling paths I usually have comments explaining what the things are which we're freeing.  e.g., "Stop accepting connections and shut down the dispatcher", "Shut down the events system", "Free arrays of resolved addresses", etc.